src/python/pants/source/source_root.py

```diff
@@ -26,18 +28,67 @@ class AllSourceRoots(Collection[SourceRoot]):
     pass
 
 
+class SourceRootPatternMatcher:
+    # We perform pattern matching against absolute paths, where "/" represents the repo root.
+    _repo_root = PurePath(os.path.sep)
+
+    def __init__(self, root_patterns: Iterable[str]):
+        self._root_patterns = list(root_patterns)
+
+    def get_patterns(self) -> Tuple[str, ...]:
+        return tuple(self._root_patterns)
+
+    def add_source_root(self, path):
+        """Add the specified fixed source root, which must be relative to the buildroot.
+
+        Useful in a limited set of circumstances, e.g., when unpacking sources from a jar with
+        unknown structure.  Tests should prefer to use dirs that match our source root patterns
+        instead of explicitly setting source roots here.
+        """
+        self._root_patterns.append(str(self._repo_root / path))
+
+    def _match_root_patterns(self, putative_root: PurePath) -> bool:
+        for pattern in self._root_patterns:
+            if putative_root.match(pattern):
+                return True
+        return False
+
+    def find_root(self, relpath: str) -> Optional[PurePath]:
+        """Return the source root for the given path, relative to the repo root."""
+        # Note: This is currently O(n) where n is the number of patterns, which
+        # we expect to be small.  We can optimize if it becomes necessary.
+        putative_root = self._repo_root / relpath
+        while putative_root != self._repo_root:
```

Comment:
Would be good to confirm that this does the right thing for something like `../../path`... ie, something that would escape the buildroot.

Reply:
Added handling for this and tests, even though you can't actually "escape the buildroot" in a meaningful sense. In the worst case, you would misidentify a source root.

Reply:
Thanks! I've seen infinite loops caused by while-loops on `parent`, so wanted to be safe =)